lib/babe/epoch.go

```diff
@@ -6,137 +6,182 @@ package babe
+	if err != nil {
+		return nil, 0, err
+	}
 
-		logger.Debugf("estimated first slot as %d based on building block 1", startSlot)
-		for i := startSlot; i < startSlot+b.epochLength; i++ {
-			_, err := b.runLottery(i, epoch)
-			if err != nil {
-				if errors.Is(err, errOverPrimarySlotThreshold) {
-					continue
-				}
-				return fmt.Errorf("error running slot lottery at slot %d: error %w", i, err)
-			}
+	has, err = b.epochState.HasConfigData(epoch)
+	if err != nil {
+		return nil, 0, err
+	}
 
-			startSlot = i
-			break
+	var cfgData *types.ConfigData
+	if has {
+		cfgData, err = b.epochState.GetConfigData(epoch)
+		if err != nil {
+			return nil, 0, err
 		}
-
-		// we are at genesis, set first slot by checking at which slot we will be able to produce block 1
-		err = b.epochState.SetFirstSlot(startSlot)
+	} else {
+		cfgData, err = b.epochState.GetLatestConfigData()
 		if err != nil {
-			return err
+			return nil, 0, err
 		}
 	}
 
-	logger.Infof("initiating epoch %d with start slot %d", epoch, startSlot)
+	threshold, err := CalculateThreshold(cfgData.C1, cfgData.C2, len(data.Authorities))
+	if err != nil {
+		return nil, 0, err
+	}
 
-	for i := startSlot; i < startSlot+b.epochLength; i++ {
-		if epoch > 0 {
-			delete(b.slotToProof, i-b.epochLength) // clear data from previous epoch
-		}
+	ed := &epochData{
+		randomness:     data.Randomness,
+		authorities:    data.Authorities,
+		authorityIndex: idx,
+		threshold:      threshold,
+	}
 
-		proof, err := b.runLottery(i, epoch)
-		if err != nil {
-			if errors.Is(err, errOverPrimarySlotThreshold) {
-				continue
-			}
-			return fmt.Errorf("error running slot lottery at slot %d: error %w", i, err)
-		}
+	startSlot, err := b.epochState.GetStartSlotForEpoch(epoch)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return ed, startSlot, nil
+}
 
-		b.slotToProof[i] = proof
-		logger.Tracef("claimed slot %d, there are now %d slots into epoch", startSlot, i-startSlot)
+func (b *Service) getLatestEpochData() (*epochData, error) {
+	var err error
+	resEpochData := &epochData{}
+
+	epochData, err := b.epochState.GetLatestEpochData()
+	if err != nil {
+		return nil, err
 	}
 
-	return nil
+	resEpochData.randomness = epochData.Randomness
+	resEpochData.authorities = epochData.Authorities
+
+	configData, err := b.epochState.GetLatestConfigData()
+	if err != nil {
+		return nil, err
+	}
+
+	resEpochData.threshold, err = CalculateThreshold(configData.C1, configData.C2, len(resEpochData.authorities))
+	if err != nil {
+		return nil, err
+	}
+
+	if !b.authority {
+		return resEpochData, nil
+	}
+
+	resEpochData.authorityIndex, err = b.getAuthorityIndex(resEpochData.authorities)
+	return resEpochData, err
 }
 
-func (b *Service) getFirstSlot(epoch uint64) (uint64, error) {
-	startSlot := getCurrentSlot(b.slotDuration)
-	for i := startSlot; i < startSlot+b.epochLength; i++ {
-		_, err := b.runLottery(i, epoch)
+func (b *Service) getFirstAuthoringSlot(epoch uint64, epochData *epochData) (uint64, error) {
```

Comment:
nit Why do we pass the epoch data by pointer? It feels it would be simpler by value 🤔